pandas/util/_test_decorators.py

```diff
@@ -232,38 +232,48 @@ def documented_fixture(fixture):
     return documented_fixture
 
 
-def check_file_leaks(func) -> Callable:
+class ResourceLeak(ContextDecorator):
     """
-    Decorate a test function to check that we are not leaking file descriptors.
+    Use psutil and ResourceWarning to identify forgotten resources.
     """
-    with file_leak_context():
-        return func
 
-
-@contextmanager
-def file_leak_context():
-    """
-    ContextManager analogue to check_file_leaks.
-    """
-    psutil = safe_import("psutil")
-    if not psutil:
-        yield
-    else:
-        proc = psutil.Process()
-        flist = proc.open_files()
-        conns = proc.connections()
-
-        yield
-
-        flist2 = proc.open_files()
-        # on some builds open_files includes file position, which we _dont_
-        #  expect to remain unchanged, so we need to compare excluding that
-        flist_ex = [(x.path, x.fd) for x in flist]
-        flist2_ex = [(x.path, x.fd) for x in flist2]
-        assert flist2_ex == flist_ex, (flist2, flist)
-
-        conns2 = proc.connections()
-        assert conns2 == conns, (conns2, conns)
+    def __enter__(self):
+        # psutil
+        self.psutil = safe_import("psutil")
+        if self.psutil:
+            self.proc = self.psutil.Process()
+            self.flist = self.proc.open_files()
+            self.conns = self.proc.connections()
+
+        # warnings
+        warnings.simplefilter("always", category=ResourceWarning)
+        self.catcher = warnings.catch_warnings(record=True)
```

Comment:
is there a better solution? This catches all warnings - they will never appear!

One (extreme?) option might be to catch all uncaught warnings for all tests and `raise` them (basically `-W error` but also for `ResourceWarning`s which are typically still ignored).